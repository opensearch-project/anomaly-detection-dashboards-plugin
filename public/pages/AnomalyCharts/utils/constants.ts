/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

export const INITIAL_ANOMALY_SUMMARY = {
  anomalyOccurrence: 0,
  minAnomalyGrade: 0.0,
  maxAnomalyGrade: 0.0,
  minConfidence: 0.0,
  maxConfidence: 0.0,
  lastAnomalyOccurrence: '',
};

export enum CHART_FIELDS {
  PLOT_TIME = 'plotTime',
  ANOMALY_GRADE = 'anomalyGrade',
  CONFIDENCE = 'confidence',
  DATA = 'data',
  AGG_INTERVAL = 'aggInterval',
  EXPECTED_VALUE = 'expectedValue',
}

export enum CHART_COLORS {
  ANOMALY_GRADE_COLOR = '#D13212',
  FEATURE_DATA_COLOR = '#16191F',
  FEATURE_COLOR = '#fcd529',
  CONFIDENCE_COLOR = '#017F75',
  LIGHT_BACKGROUND = '#FFFFFF',
  DARK_BACKGROUND = '#1D1E24',
}

export const FEATURE_CHART_THEME = [
  {
    lineSeriesStyle: {
      line: {
        strokeWidth: 2,
        visible: true,
        opacity: 0.5,
      },
      point: {
        visible: true,
        stroke: CHART_COLORS.FEATURE_DATA_COLOR,
      },
    },
  },
  {
    colors: {
      vizColors: [CHART_COLORS.FEATURE_DATA_COLOR],
    },
  },
];

export const ANOMALY_CHART_THEME = [
  {
    colors: {
      vizColors: [
        CHART_COLORS.CONFIDENCE_COLOR,
        CHART_COLORS.ANOMALY_GRADE_COLOR,
      ],
    },
  },
];

export const DATE_PICKER_QUICK_OPTIONS = [
  { start: 'now-24h', end: 'now', label: 'last 24 hours' },
  { start: 'now-7d', end: 'now', label: 'last 7 days' },
  { start: 'now-30d', end: 'now', label: 'last 30 days' },
  { start: 'now-90d', end: 'now', label: 'last 90 days' },

  { start: 'now/d', end: 'now', label: 'Today' },
  { start: 'now/w', end: 'now', label: 'Week to date' },
  { start: 'now/M', end: 'now', label: 'Month to date' },
  { start: 'now/y', end: 'now', label: 'Year to date' },
];

export enum LIVE_CHART_CONFIG {
  REFRESH_INTERVAL_IN_SECONDS = 30 * 1000, //poll anomaly result every 30 seconds
  MONITORING_INTERVALS = 60,
}

export const DEFAULT_ANOMALY_SUMMARY = {
  anomalyOccurrence: 0,
  minAnomalyGrade: 0,
  maxAnomalyGrade: 0,
  minConfidence: 0,
  maxConfidence: 0,
  lastAnomalyOccurrence: '-',
};

export const HEATMAP_CHART_Y_AXIS_WIDTH = 30;
