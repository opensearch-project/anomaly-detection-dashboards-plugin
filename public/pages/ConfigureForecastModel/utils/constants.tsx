/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import {
  ModelConfigurationFormikValues,
} from '../../ConfigureForecastModel/models/interfaces';
import { DEFAULT_SHINGLE_SIZE } from '../../../utils/constants';

// FIXME: We intentionally leave some parameters undefined to encourage customers to click 
// the "Suggest parameters" button rather than providing default values like "10 minutes" for interval
export const INITIAL_MODEL_CONFIGURATION_VALUES: ModelConfigurationFormikValues =
  {
    shingleSize: DEFAULT_SHINGLE_SIZE,
    imputationOption: undefined,
    interval: undefined,
    windowDelay: undefined,
    suggestedSeasonality: undefined,
    recencyEmphasis: undefined,
    resultIndex: undefined,
    resultIndexMinAge: 7,
    resultIndexMinSize: 51200,
    resultIndexTtl: 60,
    flattenCustomResultIndex: false,
  };

// an enum for the sparse data handling options
export enum SparseDataOptionValue {
  IGNORE = 'ignore',
  PREVIOUS_VALUE = 'previous_value',
  SET_TO_ZERO = 'set_to_zero',
  CUSTOM_VALUE = 'custom_value',
}
