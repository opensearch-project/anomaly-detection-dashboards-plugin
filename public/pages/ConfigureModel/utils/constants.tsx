/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { FEATURE_TYPE } from '../../../models/interfaces';
import {
  ModelConfigurationFormikValues,
  FeaturesFormikValues,
} from '../../ConfigureModel/models/interfaces';
import { DEFAULT_SHINGLE_SIZE } from '../../../utils/constants';

export const INITIAL_MODEL_CONFIGURATION_VALUES: ModelConfigurationFormikValues =
  {
    featureList: [],
    categoryFieldEnabled: false,
    categoryField: [],
    shingleSize: DEFAULT_SHINGLE_SIZE,
    imputationOption: undefined
  };

export const INITIAL_FEATURE_VALUES: FeaturesFormikValues = {
  featureId: '',
  featureName: '',
  featureEnabled: true,
  featureType: FEATURE_TYPE.SIMPLE,
  aggregationQuery: JSON.stringify(
    { aggregation_name: { sum: { field: 'field_name' } } },
    null,
    4
  ),
  aggregationBy: '',
  aggregationOf: [],
};

export const FEATURE_TYPES = [
  { text: 'Custom Aggregation', value: FEATURE_TYPE.CUSTOM },
  { text: 'Defined Aggregation', value: FEATURE_TYPE.SIMPLE },
];

export const FEATURE_TYPE_OPTIONS = [
  { text: 'Field value', value: FEATURE_TYPE.SIMPLE },
  { text: 'Custom expression', value: FEATURE_TYPE.CUSTOM },
];

export enum SAVE_FEATURE_OPTIONS {
  START_AD_JOB = 'start_ad_job',
  KEEP_AD_JOB_STOPPED = 'keep_ad_job_stopped',
}

export const AGGREGATION_TYPES = [
  { value: 'avg', text: 'average()' },
  { value: 'value_count', text: 'count()' },
  { value: 'sum', text: 'sum()' },
  { value: 'min', text: 'min()' },
  { value: 'max', text: 'max()' },
];

export const FEATURE_FIELDS = [
  'featureName',
  'aggregationOf',
  'aggregationBy',
  'aggregationQuery',
];

// an enum for the sparse data handling options
export enum SparseDataOptionValue {
  IGNORE = 'ignore',
  PREVIOUS_VALUE = 'previous_value',
  SET_TO_ZERO = 'set_to_zero',
  CUSTOM_VALUE = 'custom_value',
}
