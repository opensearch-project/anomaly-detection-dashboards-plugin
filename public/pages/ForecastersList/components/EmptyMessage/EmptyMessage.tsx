/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { EuiSmallButton, EuiEmptyPrompt, EuiText, EuiLink } from '@elastic/eui';
import React from 'react';
import { CreateForecasterButtons } from '../../../../components/CreateForecasterButtons/CreateForecasterButtons';
import { FORECASTER_DOCS_LINK } from '../../../../utils/constants';

const FILTER_TEXT =
  'There are no results for your search. Reset the search criteria to view forecasts.';
const EMPTY_TEXT = 'Forecasts appear here.';

interface EmptyForecasterProps {
  isFilterApplied: boolean;
  onResetFilters: () => void;
}

export const EmptyForecasterMessage = (props: EmptyForecasterProps) => (
  <EuiEmptyPrompt
    data-test-subj="emptyForecasterListMessage"
    style={{ maxWidth: '45em' }}
    body={
      <EuiText size="s">
        <p>
          {props.isFilterApplied ? FILTER_TEXT : EMPTY_TEXT}
          {!props.isFilterApplied && (
            <>
              {' '}
              <EuiLink href={`${FORECASTER_DOCS_LINK}`} target="_blank">
                Learn more
              </EuiLink>
            </>
          )}
        </p>
      </EuiText>
    }
    actions={
      props.isFilterApplied ? (
        <EuiSmallButton
          fill
          onClick={props.onResetFilters}
          data-test-subj="resetListFilters"
        >
          Reset search criteria
        </EuiSmallButton>
      ) : (
        <CreateForecasterButtons />
      )
    }
  />
);
