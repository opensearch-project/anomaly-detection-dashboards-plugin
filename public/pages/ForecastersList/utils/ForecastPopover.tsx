/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React, { useState } from 'react';
import { EuiHealth, EuiSmallButton, EuiLink, EuiSpacer, EuiText, EuiHorizontalRule, EuiCodeBlock, EuiCallOut } from '@elastic/eui';
import { ForecasterListItem } from '../../../models/interfaces';
import { FORECASTER_STATE, FORECASTER_STATE_TO_DISPLAY, isForecasterErrorState } from '../../../../server/utils/constants';
import moment from 'moment';
import { FORECASTING_FEATURE_NAME } from '../../../utils/constants';

interface ForecasterPopoverProps {
  forecaster: ForecasterListItem;
  stateColor: string;
  timeField: 'realTimeLastUpdateTime' | 'runOnceLastUpdateTime';
  message: string;
  dataSourceId?: string,
}

const buildForecasterUrl = (forecasterId: string, dataSourceId?: string) => {
  let url = `${FORECASTING_FEATURE_NAME}#/forecasters/${forecasterId}/details`;

  if (dataSourceId) {
    url += `?dataSourceId=${dataSourceId}`;
  }

  // Add a unique timestamp to force a complete page refresh
  // This ensures the ForecasterDetail component fully remounts and loads fresh data
  url += (dataSourceId ? '&' : '?') + '_t=' + Date.now();

  return url;
};

// FIXME: Cannot use EuiPopover here as it conflicts with EuiDataGrid's enlarge cell action.
// Current implementation uses a custom popover solution as a workaround.
export function ForecasterPopover(props: ForecasterPopoverProps) {
  const {
    forecaster,
    stateColor,
    timeField,
    message,
    dataSourceId,
  } = props;

  // FIXME: Attempted to implement error display in a separate modal, but encountered two major issues:
  // 1. Popover would remain on top of modal, partially hiding error message
  // 2. Attempts to close popover resulted in an empty popover still being displayed
  // Current workaround: Display error message within the same popover instead of a separate modal
  const [showFullError, setShowFullError] = useState(false);

  const forecasterUrl = buildForecasterUrl(forecaster.id, dataSourceId);

  const formattedTime = forecaster[timeField]
    ? moment(forecaster[timeField]).format('MMM DD, YYYY HH:mm:ss')
    : 'N/A';

  // If we are not showing the full error, show the standard popover content
  if (!showFullError) {
    return (
      <div style={popoverStyles}>
        <EuiHealth color={stateColor}>
          {FORECASTER_STATE_TO_DISPLAY[forecaster.curState]}
        </EuiHealth>
        {/* Override default margin-top from EuiHorizontalRule to reduce space below EuiHealth */}
        <EuiHorizontalRule margin="xs" style={{ marginTop: 0 }}/>

        <EuiText size="s">
          <p>
            {/* tabIndex={-1} prevents the link from receiving automatic focus when popover opens,
                * while maintaining clickability. Without this, EUI's accessibility features would
                * automatically focus the link, causing unwanted highlighting.
                */}
            <EuiLink href={forecasterUrl} tabIndex={-1}>
              {forecaster.name}
            </EuiLink>
            {' '}
            {message} {formattedTime}
          </p>
        </EuiText>
        <EuiSpacer size="s" />


        {forecaster.stateError && isForecasterErrorState(forecaster.curState) ? (
          <EuiSmallButton
            color="danger"
            onClick={() => setShowFullError(true)}
            fullWidth
          >
            See full error
          </EuiSmallButton>
        ) : (
          <EuiSmallButton
            fullWidth
            onClick={() => window.location.href = forecasterUrl}
          >
            View forecast
          </EuiSmallButton>
        )}
      </div>
    );
  }

  // If showFullError = true, display the entire error text
  return (
    <div style={popoverStyles}>
      <EuiHealth color={stateColor}>
        {FORECASTER_STATE_TO_DISPLAY[forecaster.curState]}
      </EuiHealth>
      <EuiHorizontalRule margin="xs" style={{ marginTop: 0 }}/>

      <EuiCallOut title="Full error details" color="danger" iconType="alert">
        <EuiText size="s">
          <p>
            <strong>{forecaster.name}</strong>
          </p>
          <p>
            {/* The date/time from your field */}
            {message} {formattedTime}
          </p>
        </EuiText>

        <EuiSpacer size="s" />
        <EuiCodeBlock
          language="bash"
          paddingSize="s"
          overflowHeight={200}
        >
          {forecaster.stateError || 'No error found'}
        </EuiCodeBlock>
      </EuiCallOut>

      <EuiSpacer size="s" />

      {/* Provide a way to go back or close */}
      <EuiSmallButton fill size="s" onClick={() => setShowFullError(false)}>
        Back
      </EuiSmallButton>
    </div>
  );
}

const popoverStyles: React.CSSProperties = {
  width: 250
};

export const getErrorModalTitle = (state: FORECASTER_STATE) => {
  switch (state) {
    case FORECASTER_STATE.INIT_ERROR:
      return 'Forecast initialization error';
    case FORECASTER_STATE.FORECAST_FAILURE:
      return 'Forecast execution error';
    case FORECASTER_STATE.INIT_TEST_FAILED:
      return 'Test initialization error';
    default:
      return 'Forecast error';
  }
};
