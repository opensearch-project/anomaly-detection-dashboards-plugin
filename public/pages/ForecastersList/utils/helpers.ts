/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import queryString from 'query-string';
import { get, } from 'lodash';
import { GetForecastersQueryParams } from '../../../../server/models/types';
import { FORECASTER_STATE, FORECASTER_STATE_DISPLAY } from '../../../../server/utils/constants';
import { Monitor } from '../../../models/interfaces';
import { ForecasterListItem } from '../../../models/interfaces';
import { FORECASTER_ACTION } from '../utils/constants';

/**
 * Parses URL query parameters from the location object's search string.
 * Currently, this function specifically looks for the 'dataSourceId' parameter.
 * It uses the 'query-string' library to parse the raw search string (e.g., "?dataSourceId=xyz")
 * into an object.
 *
 * Example:
 * If location.search is "?dataSourceId=abc&otherParam=123", this function will return
 * { dataSourceId: 'abc' }.
 * If location.search is "?someOtherParam=456" or empty, it will return
 * { dataSourceId: undefined }.
 *
 * @param location - The location object, typically provided by a router (like react-router),
 *                   containing the current URL parts. Only the 'search' property is used.
 * @param location.search - The query string part of the URL (e.g., "?param1=value1&param2=value2").
 * @returns An object conforming to GetForecastersQueryParams, containing the extracted
 *          'dataSourceId' if present, otherwise 'dataSourceId' will be undefined.
 */
export const getURLQueryParams = (location: {
  search: string;
}): GetForecastersQueryParams => {
  const {
    dataSourceId,
  } = queryString.parse(location.search) as { [key: string]: string };
  return {
    // @ts-ignore
    dataSourceId: dataSourceId === undefined ? undefined : dataSourceId,
  };
};

export const getForecasterStateOptions = () => {
  // Get unique display values from the enum
  const uniqueDisplayStates = [
    ...new Set(Object.values(FORECASTER_STATE_DISPLAY)),
  ];
  return uniqueDisplayStates.map((displayState) => ({
    label: displayState,
    text: displayState, // Using display state for text as well
  }));
};

export const getActionsForForecaster = (
  forecaster: ForecasterListItem
) => {
  switch (forecaster.curState) {
    case FORECASTER_STATE.INACTIVE_NOT_STARTED:
    case FORECASTER_STATE.INACTIVE_STOPPED:
    case FORECASTER_STATE.TEST_COMPLETE:
    case FORECASTER_STATE.FORECAST_FAILURE:
    case FORECASTER_STATE.INIT_ERROR:
    case FORECASTER_STATE.INIT_TEST_FAILED: {
      return [
        FORECASTER_ACTION.TEST,
        FORECASTER_ACTION.START,
        FORECASTER_ACTION.DELETE
      ]
    }
    case FORECASTER_STATE.AWAITING_DATA_TO_INIT:
    case FORECASTER_STATE.AWAITING_DATA_TO_RESTART:
    case FORECASTER_STATE.INIT_TEST:
    case FORECASTER_STATE.INITIALIZING_FORECAST: {
      return [
        FORECASTER_ACTION.CANCEL,
        FORECASTER_ACTION.DELETE
      ];
    }
    case FORECASTER_STATE.RUNNING: {
      return [
        FORECASTER_ACTION.STOP,
        FORECASTER_ACTION.DELETE
      ];
    }
    default:
      return [];
  }
};

export const getMonitorsForAction = (
  forecastersForAction: ForecasterListItem[],
  monitors: { [key: string]: Monitor }
) => {
  let monitorsForAction = {} as { [key: string]: Monitor };
  forecastersForAction.forEach((forecaster) => {
    const relatedMonitor = get(monitors, `${forecaster.id}.0`);
    if (relatedMonitor) {
      monitorsForAction[`${forecaster.id}`] = relatedMonitor;
    }
  });
  return monitorsForAction;
};
