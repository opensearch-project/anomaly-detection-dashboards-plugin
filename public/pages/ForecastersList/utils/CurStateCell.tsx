/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import { forecastStateToColorMap } from '../../utils/constants';
import { FORECASTER_STATE } from '../../../../server/utils/constants';
import { ForecasterListItem } from '../../../models/interfaces';
import { ForecasterPopover } from './ForecastPopover';

interface CurStateCellProps {
  forecaster: ForecasterListItem;
  dataSourceId?: string;
  isExpanded: boolean; 
  onForceCollapse?: () => void;
}

export function CurStateCell({
  forecaster,
  dataSourceId,
}: CurStateCellProps) {

  return (
    <div>
      {stateRenderFactory(forecaster, dataSourceId)}
    </div>
  );
}

function stateRenderFactory(
    forecaster: ForecasterListItem,
    dataSourceId: string | undefined,
) {

    const curState = forecaster.curState;
    const stateColor = forecastStateToColorMap.get(curState);
  
    let popoverContent;
  
    switch (curState) {
      case FORECASTER_STATE.INACTIVE_STOPPED: {
        popoverContent = <ForecasterPopover
          forecaster={forecaster}
          stateColor={stateColor || ''}
          timeField='realTimeLastUpdateTime'
          message='stopped at'
          dataSourceId={dataSourceId}
        />;
        break;
      }
  
      case FORECASTER_STATE.AWAITING_DATA_TO_INIT: {
        popoverContent = <ForecasterPopover
          forecaster={forecaster}
          stateColor={stateColor || ''}
          timeField='realTimeLastUpdateTime'
          message='has been awaiting sufficient data to initialize forecast since'
          dataSourceId={dataSourceId}
        />;
        break;
      }
  
      case FORECASTER_STATE.AWAITING_DATA_TO_RESTART: {
        popoverContent = <ForecasterPopover
          forecaster={forecaster}
          stateColor={stateColor || ''}
          timeField='realTimeLastUpdateTime'
          message='has been awaiting sufficient data to restart forecast since'
          dataSourceId={dataSourceId}
        />;
        break;
      }
  
      case FORECASTER_STATE.INIT_TEST: {
        popoverContent = <ForecasterPopover
          forecaster={forecaster}
          stateColor={stateColor || ''}
          timeField='runOnceLastUpdateTime'
          message='has been initializing test since'
          dataSourceId={dataSourceId}
        />;
        break;
      }
  
      case FORECASTER_STATE.INITIALIZING_FORECAST: {
        popoverContent = <ForecasterPopover
          forecaster={forecaster}
          stateColor={stateColor || ''}
          timeField='realTimeLastUpdateTime'
          message='initializing forecast since'
          dataSourceId={dataSourceId}
        />;
        break;
      }
  
      case FORECASTER_STATE.TEST_COMPLETE: {
        popoverContent = <ForecasterPopover
          forecaster={forecaster}
          stateColor={stateColor || ''}
          timeField='runOnceLastUpdateTime'
          message='test complete at'
          dataSourceId={dataSourceId}
        />;
        break;
      }
  
      case FORECASTER_STATE.RUNNING: {
        popoverContent = <ForecasterPopover
          forecaster={forecaster}
          stateColor={stateColor || ''}
          timeField='realTimeLastUpdateTime'
          message='has been running since'
          dataSourceId={dataSourceId}
        />;
        break;
      }
  
      case FORECASTER_STATE.INIT_ERROR:
      case FORECASTER_STATE.FORECAST_FAILURE:
      case FORECASTER_STATE.INIT_TEST_FAILED: {
        // FIXME: Using "since" instead of "at" for real-time forecast errors because:
        // 1. Real-time jobs continue running even when in error state
        // 2. Timestamp won't update on subsequent failures
        // 3. "Since" better indicates the ongoing nature of the error state
        const errorMessage = curState === FORECASTER_STATE.INIT_ERROR
          ? 'forecast initialization failed since'
          : forecaster.curState === FORECASTER_STATE.INIT_TEST_FAILED
            ? 'test initialization failed at'
            : 'forecast failed since';
        popoverContent = <ForecasterPopover
          forecaster={forecaster}
          stateColor={stateColor || ''}
          timeField={curState === FORECASTER_STATE.INIT_TEST_FAILED ? 'runOnceLastUpdateTime' : 'realTimeLastUpdateTime'}
          message={errorMessage}
          dataSourceId={dataSourceId}
        />;
        break;
      }
  
      default:
        return null;
    }
  
    return ( 
        popoverContent
    );
  }
