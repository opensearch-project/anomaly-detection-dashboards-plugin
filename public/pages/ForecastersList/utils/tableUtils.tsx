/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import {
  EuiLink,
  EuiToolTip,
  EuiHealth,
  EuiDataGridColumn,
} from '@elastic/eui';
//@ts-ignore
import moment from 'moment';
import React from 'react';
import { FORECASTING_FEATURE_NAME, } from '../../../utils/constants';
import { FORECASTER_STATE, FORECASTER_STATE_TO_DISPLAY } from '../../../../server/utils/constants';
import { forecastStateToColorMap } from '../../utils/constants';
import { CurStateCell } from './CurStateCell';

export const DEFAULT_EMPTY_DATA = '-';

export const columnStyle = {
  overflow: 'visible',
  whiteSpace: 'normal',
  wordBreak: 'break-word',
} as React.CSSProperties;

export const renderTime = (time: number) => {
  const momentTime = moment(time);
  if (time && momentTime.isValid())
    return momentTime.format('MM/DD/YYYY h:mm A');
  return DEFAULT_EMPTY_DATA;
};

function renderIndices(indices) {
  if (!indices || indices.length === 0) {
    return DEFAULT_EMPTY_DATA;
  }
  return <>{indices.join(', ')}</>;
}

export const renderState = (state: FORECASTER_STATE) => {
  return (
    //@ts-ignore
    <EuiHealth color={forecastStateToColorMap.get(state)}>{FORECASTER_STATE_TO_DISPLAY[state]}</EuiHealth>
  );
};

export function getDataGridColumns(): EuiDataGridColumn[] {
  return [
    {
      id: 'name',
      displayAsText: 'Name',
      display: (
        <EuiToolTip content="The name of the forecaster">
          <span style={columnStyle}>Name</span>
        </EuiToolTip>
      ),
      isSortable: true,
      schema: 'string',
    },
    {
      id: 'indices',
      displayAsText: 'Indices',
      display: (
        <EuiToolTip content="The index or index pattern used for the forecaster">
          <span style={columnStyle}>Indices</span>
        </EuiToolTip>
      ),
      isSortable: true,
      schema: 'string',
    },
    {
      id: 'curState',
      displayAsText: 'Status',
      display: (
        <EuiToolTip content="The current state of the forecaster">
          <span style={columnStyle}>Status</span>
        </EuiToolTip>
      ),
      isSortable: true,
      schema: 'string',
      isExpandable: true,
    },
    {
      id: 'lastUpdateTime',
      displayAsText: 'Last updated',
      display: (
        <EuiToolTip content="Time of the last update">
          <span style={columnStyle}>Last updated</span>
        </EuiToolTip>
      ),
      isSortable: true,
      schema: 'datetime', // from dataType 'date'
      defaultSortDirection: 'desc',
    },
  ];
}

export function renderCellValueFactory(
  forecasters, 
  dataSourceId?, 
  forceCollapsedRows?: Set<number>,
  forceCollapseRow?: (rowIndex: number) => void) {
  return function RenderCellValue({ rowIndex, columnId, setCellProps, isExpandable, isExpanded, isDetails }) {
    try {
      const forecaster = forecasters[rowIndex];
      if (!forecaster || rowIndex >= forecasters.length) {
        return null;
      }

      const value = forecaster[columnId];

      switch (columnId) {
        case 'name': {
          let href = `${FORECASTING_FEATURE_NAME}#/forecasters/${forecaster.id}/details`;
          
          if (dataSourceId) {
            href += `?dataSourceId=${dataSourceId}`;
          }
          
          // Add a timestamp to force a full component remount when navigating to details
          // This prevents issues with stale React refs that can cause endless loading
          // The URL parameter ensures React Router treats this as a completely fresh route,
          // properly resetting all component state including refs.
          href += (dataSourceId ? '&' : '?') + '_t=' + Date.now();
          
          return <EuiLink href={href}>{value}</EuiLink>;
        }

        case 'indices':
          return renderIndices(value);

        case 'curState':{
          // Check if the parent said "force collapse this row"
          const forciblyCollapsed = forceCollapsedRows?.has(rowIndex);

          // If the cell is not expanded, OR it's forcibly collapsed, just show the normal cell
          if (!isExpanded || forciblyCollapsed || value === FORECASTER_STATE.INACTIVE_NOT_STARTED) {
            return renderState(value);
          }

          // Otherwise, render CurStateCell
          return (
            <CurStateCell
              forecaster={forecaster}
              dataSourceId={dataSourceId}
              isExpanded={isExpanded}
              onForceCollapse={() => forceCollapseRow?.(rowIndex)} // pass rowIndex
            />
          );
        }

        case 'lastUpdateTime':
          return renderTime(value);

        default:
          return <>{value}</>;
      }
    } catch (error) {
      console.error('Error rendering cell value:', error);
      return DEFAULT_EMPTY_DATA;
    }
  };
}
