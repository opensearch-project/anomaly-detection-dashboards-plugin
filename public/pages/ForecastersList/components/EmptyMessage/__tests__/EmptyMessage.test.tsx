/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { EmptyForecasterMessage } from '../EmptyMessage';
import { BASE_DOCS_LINK } from '../../../../../utils/constants';

jest.mock('../../../../../services', () => ({
  ...jest.requireActual('../../../../../services'),
  getDataSourceEnabled: () => ({
    enabled: false  
  })
}));

describe('<EmptyForecasterMessage /> spec', () => {
  beforeAll(() => {
    Object.defineProperty(window, 'location', {
      value: {
        href: 'http://test.com',
        pathname: '/',
        search: '',
        hash: '',
      },
      writable: true
    });
  });

  describe('Empty results', () => {
    test('renders component with empty message and learn more link', () => {
      const history = createMemoryHistory(); 

      const { container, getByText } = render(
        <Router history={history}>
          <EmptyForecasterMessage
            isFilterApplied={false}
            onResetFilters={jest.fn()}
          />
        </Router>
      );
      expect(container.firstChild).toMatchSnapshot();
      expect(getByText('Forecasts appear here.')).toBeInTheDocument();
      expect(getByText('Learn more')).toBeInTheDocument();
      expect(getByText('Learn more').closest('a')).toHaveAttribute('href', `${BASE_DOCS_LINK}/forecast`);
    });
  });

  describe('Filters results message', () => {
    test('renders component no result for filters message without learn more link', () => {
      const { container, queryByText } = render(
        <EmptyForecasterMessage
          isFilterApplied={true}
          onResetFilters={jest.fn()}
        />
      );
      expect(container.firstChild).toMatchSnapshot();
      expect(queryByText('Learn more')).not.toBeInTheDocument();
    });

    test('resets filters when click on reset filters', () => {
      const handleResetFilters = jest.fn();
      const { getByTestId } = render(
        <EmptyForecasterMessage
          isFilterApplied={true}
          onResetFilters={handleResetFilters}
        />
      );
      fireEvent.click(getByTestId('resetListFilters'));
      expect(handleResetFilters).toHaveBeenCalled();
      expect(handleResetFilters).toHaveBeenCalledTimes(1);
    });
  });
});
