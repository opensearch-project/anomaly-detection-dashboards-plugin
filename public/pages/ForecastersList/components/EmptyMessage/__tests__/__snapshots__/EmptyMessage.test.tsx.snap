// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmptyDetectorMessage /> spec Empty results renders component with empty message 1`] = `
<div
  class="euiEmptyPrompt"
  data-test-subj="emptyDetectorListMessage"
  style="max-width: 45em;"
>
  <span
    class="euiTextColor euiTextColor--subdued"
  >
    <div
      class="euiText euiText--medium"
    >
      <div
        class="euiText euiText--small"
      >
        <p>
          A detector is an individual anomaly detection task. You can create multiple detectors, and all the detectors can run simultaneously, with each analyzing data from different sources. Create an anomaly detector to get started.
        </p>
      </div>
    </div>
  </span>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--justifyContentCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
  >
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <a
        class="EuiSmallButton EuiSmallButton--primary EuiSmallButton--small"
        data-test-subj="sampleDetectorButton"
        href="anomaly-detection-dashboards#/overview?"
        rel="noreferrer"
        style="width: 200px;"
      >
        <span
          class="EuiSmallButtonContent EuiSmallButton__content"
        >
          <span
            class="EuiSmallButton__text"
          >
            Try a sample detector
          </span>
        </span>
      </a>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <a
        class="EuiSmallButton EuiSmallButton--primary EuiSmallButton--small EuiSmallButton--fill"
        data-test-subj="createDetectorButton"
        href="anomaly-detection-dashboards#/create-ad/?"
        rel="noreferrer"
        style="width: 200px;"
      >
        <span
          class="EuiSmallButtonContent EuiSmallButton__content"
        >
          <span
            class="EuiSmallButton__text"
          >
            Create detector
          </span>
        </span>
      </a>
    </div>
  </div>
</div>
`;

exports[`<EmptyDetectorMessage /> spec Filters results message renders component no result for filters message 1`] = `
<div
  class="euiEmptyPrompt"
  data-test-subj="emptyDetectorListMessage"
  style="max-width: 45em;"
>
  <span
    class="euiTextColor euiTextColor--subdued"
  >
    <div
      class="euiText euiText--medium"
    >
      <div
        class="euiText euiText--small"
      >
        <p>
          There are no detectors matching your applied filters. Reset your filters to view all detectors.
        </p>
      </div>
    </div>
  </span>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <button
    class="EuiSmallButton EuiSmallButton--primary EuiSmallButton--small EuiSmallButton--fill"
    data-test-subj="resetListFilters"
    type="button"
  >
    <span
      class="EuiSmallButtonContent EuiSmallButton__content"
    >
      <span
        class="EuiSmallButton__text"
      >
        Reset filters
      </span>
    </span>
  </button>
</div>
`;
